src/mds/Locker.cc

```diff
@@ -4711,19 +4728,29 @@ void Locker::mark_updated_scatterlock(ScatterLock *lock)
  *
  * Return Value: true if c is put into some wait queue, false otherwise
  */
-bool Locker::scatter_nudge(ScatterLock *lock, MDSContext *c, bool under_rstat_propagation)
+bool Locker::scatter_nudge(ScatterLock *lock, MDSContext *c, mds_rank_t remote_rstatflush_asker, bool local_rstatflush)
 {
   CInode *p = static_cast<CInode *>(lock->get_parent());
 
   if (p->is_frozen() || p->is_freezing()) {
     dout(10) << "scatter_nudge waiting for unfreeze on " << *p << dendl;
-    if (c) {
+    if (c && remote_rstatflush_asker == MDS_RANK_NONE) {
       p->add_waiter(MDSCacheObject::WAIT_UNFREEZE, c);
```

Comment:
please explain why  'adding a unfreeze waiter' does not work, you need to modify the unfreeze_tree code

Reply:
Hi, I thought there could exists chances that when the unfreeze waiter is invoked, the inode that it waited on has already been trimmed out of the memory if, after unfreeze it's not auth anymore, which means the scatter_nudge for that inode's nestlock won't be able to start.

So I added a "asker" to send out a nudge message further to the new auth, or do the nudge if the inode is still auth.

Am I right? Thanks:-)

Reply:
how about adding some code to Migrator::decode_import_inode(). the new auth mds should also contain a MDRequest (for propagating rstat)

Reply:
Yes, I'm trying that now:-)

Reply:
@ukernel Sorry, I don't quite follow you about this.
The problem is like this, say there is a directory /a/b/c/d/e/f/g/h:

1. /a is on mds.X
2. /a/b/c/d is on mds.Y
3. a rstat flush op want to flush rstat of subtree /a
4. mds.X send a slave request to mds.Y to first flush rstat flush subtree /a/b/c/d, the last step of which is to nudge the inode "/a/b/c/d"
5. mds.Y sends a LOCK_AC_NUDGE to mds.X  to nudge inode of "/a/b/c/d", but at that time subtree /a/b is being migrated to mds.Z, which would lead to the nudge of "/a/b/c/d" suspended.

As you can see, as for now, the processing of LOCK_AC_NUDGE doesn't add any waiter for the unfreeze event, which means when the migration of subtree /a/b finishes, there won't be a nudge of inode "/a/b/c/d", so the slave request that mds.X send out to mds.Y to flush rstat of "/a/b/c/d" will wait on the nudge of inode "/a/b/c/d" infinitely if there is no further action. And even if we add a waiter for the unfreeze event when processing the LOCK_AC_NUDGE message, we may encounter the problem that I mentioned above.

I think if we want to solve this problem by adding code in Migrator::decode_import_inode(), the only solution is to add a waiter that will unconditionally trigger a scatter_nudge of inode "/a/b/c/d"  when the migration finishes in Migrator::decode_import_inode(). Is this what you are suggesting? Thank you:-)

Reply:
For the case above, mds.x will send OP_PROPAGATERSTATS to mds.z. I don't see why it needs special handling.

Reply:
@ukernel After mds.X instructed mds.Y to flush its rstats, it may wait on that instruction to return. And at the time mds.X instructed mds.Y, the subtree /a/b is still on mds.X, so it wouldn't send OP_PROPAGATERSTATS. So the situation is mds.X wait on mds.Y, and mds.Y is waiting on a nudge to finish, while, unfortunately, the nudge is lost as the lock it's trying to nudge is being migrated.

Reply:
@ukernel If we make sure scatter_nudge requeue lock to updated_scatterlocks whenever they are dirty before the scatter_nudge returns, later scatter_tick can nudge that lock again, which would prevent this forever waiting. Do you think this is an acceptable solution? Thanks:-)

Reply:
I prefer the tick approach.

Reply:
@ukernel Ok:-)

Reply:
@ukernel Hi, by the way, please take a look at [29721](https://github.com/ceph/ceph/pull/29721) when you are free:-) It's also needed for this PR to run correctly:-)